lib/spack/spack/preferred_packages.py

```diff
@@ -162,8 +162,8 @@ def spec_preferred_variants(self, pkgname):
         """Return a VariantMap of preferred variants and their values"""
         variants = self.preferred.get(pkgname, {}).get('variants', '')
         if not isinstance(variants, basestring):
-            variants = "".join(variants)
-        return spack.spec.Spec(pkgname + variants).variants
+            variants = " ".join(variants)
```

Comment:
Hmm, this might be a problem. `hdf5-mpi` won't work because it is a valid package name, it has to be `hdf5 -mpi`. And `hdf5 ~mpi` won't work because mpi could theoretically be a username, it has to be `hdf5~mpi`.

Also, is variants guaranteed to be a list, or can it be a string?

I would leave it the way things were and simply replace `-` with `~` in variants.